proofs/tree.go

```diff
@@ -317,19 +321,39 @@ func (doctree *DocumentTree) AddLeavesFromDocument(document proto.Message) (err
 	if doctree.hash == nil {
 		return fmt.Errorf("hash is not set")
 	}
-	var getSalt GetSalt
-	if doctree.getSalt != nil {
-		getSalt = doctree.getSalt
+	var salts Salts
+	if doctree.salts != nil {
+		salts = doctree.salts
 	} else {
-		getSalt = defaultGetSalt(doctree.salts)
+		var err error
+		salts, err = defaultGetSalt(document)
+		if (err != nil){
+			return err
+		}
 	}
-	leaves, err := FlattenMessage(document, getSalt, doctree.saltsLengthSuffix, doctree.hash, doctree.compactProperties, doctree.parentPrefix)
+	leaves, err := FlattenMessage(document, salts, doctree.readablePropertyLengthSuffix, doctree.hash, doctree.compactProperties, doctree.parentPrefix)
 	if err != nil {
 		return err
 	}
 	return doctree.AddLeaves(leaves)
 }
 
+func fillBackSalts(message proto.Message, saltsSlice []*proofspb.Salt)(err error){
+	value := reflect.ValueOf(message).Elem().FieldByName(SaltsFieldName)
+	if value == reflect.ValueOf(nil) {
+		return errors.New("Cannot find salts field in message")
+	}
+	value.Set(reflect.ValueOf(saltsSlice))
+	return nil
+}
+
+func getSaltsFromMessage(message proto.Message) (salts []*proofspb.Salt, err error){
+	field := reflect.ValueOf(message).Elem().FieldByName(SaltsFieldName)
+	if field == reflect.ValueOf(nil) {
+		return nil,  errors.New("Cannot find salts field in message")
```

Comment:
Is it possible to error out only if salts is not a field? If salts is a field then it could be automatically populated with `[]proofspb.Salt{}`

Reply:
not catch up with you.
If a message is fed to precise, if user does not make any operation on salts field, what client get is salts which are all generated randomly

Reply:
`if field == reflect.ValueOf(nil)` means dont contain salts field instead of  nil salts field

Reply:
You are correct.